### ISSUE TITLE
Cloudpickle fails to preserve torch gradients

### ISSUE THREAD

**Issue by cjao** [AgnostiqHQ/covalent#649]
We use cloudpickle everywhere to serialize and deserialize data. Unfortunately Cloudpickle seems to lose crucial properties of Torch Tensors, such as gradients:
```
import torch
from dask.distributed.protocol import serialize, deserialize

def loss(x):
    l = torch.sum(x**2)
    return l

x = torch.tensor([2.0], requires_grad=True)
cost = loss(x)

cost.backward()

print(x, type(x), x.grad)
```
```
tensor([2.], requires_grad=True) <class 'torch.Tensor'> tensor([4.])
```
Now let's try pickling and unpickling:
```
import cloudpickle as pickle

y = pickle.loads(pickle.dumps(x))
print(x, type(x), x.grad)
print(y, type(y), y.grad)
```
```
tensor([2.], requires_grad=True) <class 'torch.Tensor'> tensor([4.])
tensor([2.], requires_grad=True) <class 'torch.Tensor'> None
```

**Comment by cjao** [comment]
Dask has special handling for Torch Tensors and other popular data structures: https://github.com/dask/distributed/tree/177dfb891089cc872bab34fb8369d75cae13bc0a/distributed/protocol

We may need to extend cloudpickle similarly. See also [this issue](https://github.com/dask/distributed/issues/2581)

**Comment by kessler-frost** [comment]
Similar issue:
https://github.com/Qiskit/qiskit-aer/issues/1537

**Comment by cjao** [comment]
Upstream [issue](https://github.com/pytorch/pytorch/issues/81186) and [proposed patch](https://github.com/pytorch/pytorch/pull/81188).

The above snippet seems to work with the patch:
```
tensor([2.], requires_grad=True) <class 'torch.Tensor'> tensor([4.])
tensor([2.], requires_grad=True) <class 'torch.Tensor'> tensor([4.])
```
